### PR TITLE
new(modern_bpf): add support for `process` family syscalls

### PR DESCRIPTION
**Pull request by Andreagit97**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area driver-modern-bpf

/area libpman

/area tests

**Does this PR require a change in the driver versions?**

**What this PR does / why we need it**:

This PR is part of a series https://github.com/falcosecurity/libs/issues/513, the final aim is to support the most important syscalls also in the new probe. This PR introduces:

* `execve`
* `execveat`
* `clone`
* `clone3`
* `fork`
* `vfork`

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(modern_bpf): add support for `process` family syscalls
```

### CODE REVIEW

**Comment by Andreagit97** [comment]
This PR and https://github.com/falcosecurity/libs/pull/544 are blocking for the last syscalls so I would focus the reviewers' attention on them if possible :pray:

**Review by leogr — DISMISSED** [review]

**Comment by Andreagit97** [comment]
Since the behavior of all these syscalls is very similar probably is better to group them all together. Sorry for the huge PR, but it's almost the same code for every syscall with just little differences

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/clone.bpf.c [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	 * - `aarch64` uses `CONFIG_CLONE_BACKWARDS`
+	 * - `s390x` uses `CONFIG_CLONE_BACKWARDS2`
+	 * - `x86_64` uses the actual version without `BACKWARDS` configs.
+	 */
+#ifdef __TARGET_ARCH_s390
+	/* In `s390` architectures `clone_flags` are the second param. */
+	unsigned long flags = extract__syscall_argument(regs, 1);
+#else
+	unsigned long flags = extract__syscall_argument(regs, 0);
+#endif
```
Comment (Andreagit97) [inline driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/clone.bpf.c:191]: We need to fix this also in the `kmod`!

File: test/modern_bpf/test_suites/syscall_enter_suite/vfork_e.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/* The vfork() function has the same effect as fork(2), except that the
+ * behavior is undefined if the process created by vfork() either modifies
+ * any data other than a variable of type pid_t used to store the return
+ * value from vfork(), or returns from the function in which vfork() was
+ * called, or calls any other function before successfully calling _exit(2)
+ * or one of the exec(3) family of functions.
+ *
+ * For this reason right now we are not able to call the `vfork()` without
+ * a segmentation fault...
+ */
```
Comment (Andreagit97) [inline test/modern_bpf/test_suites/syscall_enter_suite/vfork_e.cpp:10]: Unfortunately, I'm not able to run any tests for the `vfork` due to its strange behavior :/ It shouldn't be a big issue since in kernel `>=5.8` `vfork` shouldn't be used anymore.

**Review by FedeDP — DISMISSED** [review]
/approve

**Review by leogr — APPROVED** [review]

**Review by FedeDP — APPROVED** [review]
/approve

**Review by leogr — APPROVED** [review]

File: driver/modern_bpf/helpers/extract/extract_from_kernel.h [hunk 3]
```diff
@@ -436,3 +436,24 @@ static __always_inline unsigned long extract__vm_swap(struct mm_struct *mm)
...
+// TTY EXTRACTION
+////////////////////////
+
+/**
+ * @brief Extract encoded tty
+ *
+ * @param task pointer to task_struct.
+ * @return encoded tty number
+ */
+static __always_inline u32 exctract__tty(struct task_struct *task)
```
Comment (hbrueckner) [inline driver/modern_bpf/helpers/extract/extract_from_kernel.h:450]: typo: `exctract__tty`... should be `extract__tty`

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	}
+	else
+	{
+		/* Parameter 16: env (type: PT_CHARBUFARRAY) */
+		unsigned long envp = extract__syscall_argument(regs, 2);
+		auxmap__store_multiple_charbufs_param_from_array(auxmap, envp, 0, USER);
+	}
+
+	/* Parameter 17: tty (type: PT_INT32) */
+	u32 tty = exctract__tty(task);
```
Comment (hbrueckner) [inline driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c:216]: typo: see comment on helper.

**Review by hbrueckner — COMMENTED** [review]
Hi @Andreagit97,
here is first round of my review comments. I will continue to review the others as well. Thanks.

File: test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	int tty = 0;
+	unsigned long minflt = 0;
+	unsigned long majflt = 0;
+	pid_t ppid = 0; /* The PID of the parent of this process. */
+	pid_t pgid = 0; /* The process group ID of the process. */
+
+	/* we could get the filename of the executable (`comm`) from proc, but it is returned
+	 * in parentheses, so for example "(bpf_test)", so we prefer to use our macro `TEST_EXECUTABLE_NAME`
```
Comment (hbrueckner) [inline test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp]: It could be possible to use `"(%s)"` fmt to read `comm`, however, `%s` will include `)` that needs to be corrected.

**Review by hbrueckner — COMMENTED** [review]
Hi @Andreagit97 , I went through the rest of the PR and they look very nice.

File: driver/modern_bpf/helpers/extract/extract_from_kernel.h [hunk 6]
```diff
@@ -195,3 +195,36 @@ static __always_inline u64 extract__capability(struct task_struct *task, enum ca
 
 	return capabilities_to_scap(((unsigned long)cap_struct.cap[1] << 32) | cap_struct.cap[0]);
 }
+
+///////////////////////////
+// CHARBUF EXTRACION
+///////////////////////////
+
+/**
+ * @brief Extract a specif charbuf pointer from an array of charbuf pointers
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply driver/modern_bpf/helpers/extract/extract_from_kernel.h:218]: ty

File: driver/modern_bpf/helpers/extract/extract_from_kernel.h (hunk 3 above)
Comment (Andreagit97) [reply driver/modern_bpf/helpers/extract/extract_from_kernel.h:450]: ty

File: test/modern_bpf/event_class/event_class.h [hunk 7]
```diff
@@ -49,6 +49,11 @@ enum direction
 #define EXIT_EVENT 0
 #define ENTER_EVENT 1
 
+/* NOTE: if we change the name of this executable
+ * we have to change also this string!
+ */
+#define TEST_EXECUTABLE_NAME "bpf_test"
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply test/modern_bpf/event_class/event_class.h:58]: sounds good, ty

File: test/modern_bpf/test_suites/syscall_enter_suite/execve_e.cpp [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+TEST(SyscallEnter, execveE)
+{
+	auto evt_test = new event_test(__NR_execve, ENTER_EVENT);
+
+	evt_test->enable_capture();
+
+	/*=============================== TRIGGER SYSCALL  ===========================*/
+
+	char pathname[] = "//**null-file-path**//";
+	const char *newargv = NULL;
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply test/modern_bpf/test_suites/syscall_enter_suite/execve_e.cpp:13]: makes sense, ty

File: test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp (hunk 5 above)
Comment (Andreagit97) [reply test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp]: yeah, not sure what is the best way to do it :thinking:

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execveat.bpf.c [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+	auxmap__store_s64_param(auxmap, ret);
+
+	struct task_struct *task = get_current_task();
+
+	/* This is a charbuf pointer array.
+	 * Every element of `argv` array is a pointer to a charbuf.
+	 * Here the first pointer points to `exe` param while all
+	 * the others point to the different args.
+	 *
+	 * Please note: this bpf program is called only if the `execveat` fails
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execveat.bpf.c:78]: Yeah you can find more context in this PR https://github.com/falcosecurity/libs/pull/141

TL;DR;
`execveat` is a wrapper of `execve`, when the call succeeds the event returned is simply an `execve` exit event. Only in case of syscall failure, we can observe an `execveat` exit event...

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/clone.bpf.c [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+	auxmap__store_s64_param(auxmap, ret);
+
+	struct task_struct *task = get_current_task();
+
+	/* We can extract `exe` (Parameter 2) and `args`(Parameter 3) only if the
+	 * syscall doesn't fail. Otherwise, they will send empty parameters.
+	 */
+	if(ret >= 0)
+	{
+		unsigned long arg_start_pointer = 0;
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/clone.bpf.c:65]: Yeah after the first scaffolding I think we need to group some part of the code behind helpers

File: test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp (hunk 5 above)
Comment (hbrueckner) [reply test/modern_bpf/test_suites/syscall_exit_suite/execve_x.cpp]: I also do not have a strong opinion.. so maybe, let's leave it as for now.

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execveat.bpf.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (hbrueckner) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execveat.bpf.c:78]: understood.. thanks for the background on it!
